### PR TITLE
[Driver][BoundsSafety] Add -fexperimental-bounds-safety flag

### PR DESCRIPTION
**Pull request by rapidsna**
-fexperimental-bounds-safety is an experimental flag for -fbounds-safety, which is a bounds-safety extension for C. -fbounds-safety will require substantial changes across the Clang codebase. So we introduce this experimental flag is to gate our incremental patches until we push the essential functionality of the extension.

-fexperimental-bounds-safety currently doesn't do anything but reporting an error when the flag is used with an unsupported source language (currently only supports C).

### CODE REVIEW

**Comment by rapidsna** [comment]
I'll push BoundsSafety.rst documentation in a separate PR.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (delcypher) [reply clang/lib/Frontend/CompilerInvocation.cpp:3855]: I think this is supposed to happen in `CompilerInvocationBase::GenerateLangArgs(...)`

**Review by delcypher — CHANGES_REQUESTED** [review]
In general looks good but there are some minor changes I'd like made.

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td]: Done!

Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td]: Done!

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp:3629]: Introduced a warning for ASM. I don't think LLVM IR/Unknown should be reachable when parsing the language arguments so I now treat them unreachable.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp:3630]: Done!

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp]: I think you're right. Removed the change.

File: clang/test/BoundsSafety/Driver/driver.c
Comment (rapidsna) [reply clang/test/BoundsSafety/Driver/driver.c]: Done!

File: clang/test/BoundsSafety/Frontend/ignored_for_asm.s
Comment (delcypher) [inline clang/test/BoundsSafety/Frontend/ignored_for_asm.s:5]: @rapidsna  Is some code missing from this PR? I see you now have a test but I can't find the code change that causes this warning to be emitted in the PR.

File: clang/test/BoundsSafety/Frontend/ignored_for_ir.ll
Comment (delcypher) [inline clang/test/BoundsSafety/Frontend/ignored_for_ir.ll:3]: It might be worth having a small comment here explaining why you expect these diagnostics to not fire.

File: clang/test/BoundsSafety/Frontend/ignored_for_asm.s
Comment (rapidsna) [reply clang/test/BoundsSafety/Frontend/ignored_for_asm.s:5]: Done!

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (delcypher) [inline clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: Nit. In this PR the flag is `-fbounds-safety-experimental` not `-fbounds-safety`.

**Review by delcypher — APPROVED** [review]
@rapidsna LGTM other than the nits.

**Comment by tbaederr** [comment]
The other experimental flags I see start with experimental, this one ends with it. Why isn't this called `-fexerimental-bounds-safety`?

**Comment by rapidsna** [comment]
> The other experimental flags I see start with experimental, this one ends with it. Why isn't this called `-fexerimental-bounds-safety`?

Oh, I can see most of them start with `-fexperimental`, not everything though. I can fix this. Is there a formal convention?

File: clang/test/BoundsSafety/Frontend/only_c_is_supported.c
Comment (nickdesaulniers) [inline clang/test/BoundsSafety/Frontend/only_c_is_supported.c:19]: why not test `-x ir` and `-x assembler` in the same test file?

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (nickdesaulniers) [inline clang/lib/Frontend/CompilerInvocation.cpp:3630]: What happens if I'm compiling `-x assembler-with-cpp` (i.e. assembler that needs to be preprocessed, `.S` files)?

Does this error?

What if I'm using the preprocessor to detect if `-fbounds-safety` is supported?

```
#if __has_attribute(bidi_whatever)
mov x0, #42
#else
#error "oh no"
#endif
```
(or perhaps `__has_feature(bounds_safety)`)

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (MaskRay) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: For assembly input, we generally ignore handling the option and leave a warning https://maskray.me/blog/2023-08-25-clang-wunused-command-line-argument#assembly-files

```
% fclang -fsanitize=address -fpatchable-function-entry=2 -c a.s
clang: warning: argument unused during compilation: '-fsanitize=address' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-fpatchable-function-entry=2' [-Wunused-command-line-argument]
```

**Review by MaskRay — CHANGES_REQUESTED** [review]

File: clang/test/BoundsSafety/Driver/driver.c
Comment (MaskRay) [reply clang/test/BoundsSafety/Driver/driver.c:1]: `clang/test/Driver/clang_f_opts.c` mixes a lot of options. Newer `-fxxx` tests can go to a dedicated `clang/test/Driver/fxxx.c`

File: clang/test/BoundsSafety/Frontend/only_c_is_supported.c
Comment (MaskRay) [inline clang/test/BoundsSafety/Frontend/only_c_is_supported.c:2]: We use a terse style and use a blank line to group different sets of tests. Here the blank line use is unnecessary and harms readability IMO.

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: @MaskRay Thank you! I can use `warning: argument unused during compilation` for the driver. Do you know what is the convention for unused flags in cc1?

File: clang/test/BoundsSafety/Driver/driver.c
Comment (rapidsna) [reply clang/test/BoundsSafety/Driver/driver.c:1]: @nickdesaulniers I was hoping that we could keep -fbounds-safety tests under a separate folder clang/test/BoundsSafety in order to avoid them from getting mixed up with other tests. And make it easier to run bounds safety related tests only. But I can see this is not what other extensions do so I can move them to the existing layout.

File: clang/test/BoundsSafety/Frontend/only_c_is_supported.c
Comment (rapidsna) [reply clang/test/BoundsSafety/Frontend/only_c_is_supported.c:19]: I think at least `-x ir` had to be a separate file because how it recognize comments (`//` vs `;`). I'll double check.

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: @nickdesaulniers They are separate because the languages (C++/Obj-C/etc) we want to make an error until we have a proper support vs a warning for assembly (or maybe others?) that are meant to be ignored. For the latter I think we can use `warning: argument unused during compilation` as @MaskRay suggested.

File: clang/test/BoundsSafety/Driver/driver.c
Comment (delcypher) [reply clang/test/BoundsSafety/Driver/driver.c:1]: @nickdesaulniers As @rapidsna said there are multiple reasons:

1. Provides a very convenient way to run **only `-fbounds-safety` tests**.
2. More likely to avoid merge conflicts with our internal code.

Out of those two I think `(1.)` is the bigger problem because it is something I rely on for local development and I think merge conflicts are still unlikely even if we put tests under their proper location under `clang/test` because we mostly created new tests and tried not to touch existing tests. If `lit` had a way of labelling tests and only running tests with a particular label then the location of the tests would be irrelevant and I would be happy for the tests to live in their normal place under `clang/test/` (e.g. `clang/test/Driver`). Unfortunately AFAIK that feature doesn't exist in lit.

To workaround this would this kind of layout be acceptable?

```
test/Driver/BoundsSafety/...
test/Frontend/BoundsSafety/...
test/Sema/BoundsSafety/...
...
```

That way the tests are conforming to Clang's test layout but its still possible to run **only the -fbounds-safety** tests by doing something like:

```
$ bin/llvm-lit -vs tools/clang/test/{Driver,Frontend,Sema}/BoundsSafety
```

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (MaskRay) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: https://maskray.me/blog/2023-08-25-clang-wunused-command-line-argument has some notes. If you don't call functions like `getLastArg` that `claim` the option, there will be a `warning: argument unused during compilation: '-fsanitize=address' [-Wunused-command-line-argument]` for assembly input.

File: clang/test/BoundsSafety/Driver/driver.c
Comment (MaskRay) [reply clang/test/BoundsSafety/Driver/driver.c:1]: Test subdirectories are added conservatively. The main use cases are target-specific tests and flang. `Driver/XRay` is somehow an unneeded use that you can ignore.
```
% echo {CodeGen,Sema,Driver}/*(/)
CodeGen/aarch64_neon_sve_bridge_intrinsics CodeGen/aarch64-sme-intrinsics CodeGen/aarch64-sve2-intrinsics CodeGen/aarch64-sve2p1-intrinsics CodeGen/aarch64-sve-intrinsics CodeGen/arc CodeGen/arm-mve-intrinsics CodeGen/assignment-tracking CodeGen/avr CodeGen/CSKY CodeGen/Inputs CodeGen/LoongArch CodeGen/PowerPC CodeGen/RISCV CodeGen/SystemZ CodeGen/VE CodeGen/WebAssembly CodeGen/X86 Sema/aarch64-sme-intrinsics Sema/aarch64-sve2-intrinsics Sema/aarch64-sve2p1-intrinsics Sema/aarch64-sve-intrinsics Sema/Inputs Driver/flang Driver/Inputs Driver/XRay
```

FWIW you can do

* `path/bin/llvm-lit clang/test/Driver/bounds-safety-*`
* `LIT_OPTS=--filter=bounds-safety path/bin/llvm-lit clang/test/Driver`

and even create an alias for `LIT_OPTS=--filter=bounds-safety `

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp:3630]: Right, this is something we could potentially allow in future if we see actual use cases.

Until we support this, `__has_feature(bounds_safety)` or such will return `false` and we should have the compiler report the warning that `-fbounds-safety` is ignored for assemblers.

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (delcypher) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: @MaskRay So we aren't explicit using `getLastArg` but it looks like we are calling it indirectly from `addOptInFlag(...)` which I think means we are unconditionally claiming the option.  In order to make your suggestion work we'd have to conditionally call `addOptInFlag` in the driver based on the selected input language but I don't think that's going to work because no `LangOptions` object exists for us query at that point in the driver AFAICT.

Also future patches we will upstream will add additional `Args.getLastArg(options::OPT_fbounds_safety_experimental, ...)` calls in the driver which means the option will get claimed there too which will break any reliance on the option not being claimed that might be implemented in this PR.

This warning is really a frontend warning and not a driver warning so I think we should use the frontend warning that @rapidsna introduced in this PR.

Comment (MaskRay) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: Conventionally the language compatibility checking and other checking is performed in Driver, not in Frontend.. If you move the language check to Driver, the diagnostic will be natural since clang integrated assembler uses `ClangAs` instead of `Clang`.

Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: @MaskRay Thanks. I still need the error for `-fbounds-safety` not being supported in C++/Obj-C/etc in the frontend because it needs to be an error for `cc1`.

I will change it so that the unused warning for assembly to be handled in the driver like the rest of the unused options for assembly.

File: clang/test/BoundsSafety/Driver/driver.c
Comment (nickdesaulniers) [reply clang/test/BoundsSafety/Driver/driver.c:1]: > And make it easier to run bounds safety related tests only.

> Provides a very convenient way to run only -fbounds-safety tests.

Then give each test a similar function name prefix, and then you can do `llvm-lit -vv llvm/test/Driver/fbounds-safety-*`, as @MaskRay said.  Don't start creating new directory structures that don't exist for every other feature implemented by the compiler.

File: clang/test/BoundsSafety/Frontend/only_c_is_supported.c
Comment (nickdesaulniers) [reply clang/test/BoundsSafety/Frontend/only_c_is_supported.c:19]: ah, makes sense, consider adding that as a comment

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: > Conventionally the language compatibility checking and other checking is performed in Driver, not in Frontend.. If you move the language check to Driver, the diagnostic will be natural since clang integrated assembler uses `ClangAs` instead of `Clang`.

@MaskRay It seems both `Clang` and `ClangAs` are invoked for the `clang` command with an assembly input (i.e., `ClangAs` is invoked after `Clang`) so it seems to me that most options are already `claimed` in `Clang` and so `warning: argument unused during compilation` doesn't seem to fire for `clang` in most cases including `-fsanitize=address`.

Also, we need to have the input language (`InputKind`) to report the right diagnostics so the frontend still seems like the right place to handle this.

Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: @MaskRay Sorry, it seems we can still do it in Driver with `clang::driver::types`. I'm going to make a change to report the "unused" warning from the driver, and report the "unsupported language" error in the frontend. Does it sound okay to you?

Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td:337]: > @MaskRay It seems both `Clang` and `ClangAs` are invoked for the `clang` command with an assembly input (i.e., `ClangAs` is invoked after `Clang`) so it seems to me that most options are already `claimed` in `Clang` and so `warning: argument unused during compilation` doesn't seem to fire for `clang` in most cases including `-fsanitize=address`.

I just confirmed that this is happening when the input file format is `.s` and no `-x` is provided from the command line. If we do `-x assembler` then only `ClangAs` is launched and the unused option warning fires as expected.

File: clang/test/BoundsSafety/Driver/driver.c
Comment (rapidsna) [reply clang/test/BoundsSafety/Driver/driver.c]: @nickdesaulniers @MaskRay Thank you! I removed the new directory and moved the tests to conform to the existing layout.

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td]: @nickdesaulniers I managed to move -x ir test to the `.c` test. Not sure why but adding `-###` option seemed to make it work.

**Comment by rapidsna** [comment]
> The other experimental flags I see start with experimental, this one ends with it. Why isn't this called `-fexerimental-bounds-safety`?

@tbaederr Thank you! I just renamed the flag to `-fexperimental-bounds-safety`.

File: clang/test/BoundsSafety/Frontend/only_c_is_supported.c
Comment (rapidsna) [reply clang/test/BoundsSafety/Frontend/only_c_is_supported.c]: @MaskRay Thank you! I removed the blank lines.

File: clang/include/clang/Basic/DiagnosticFrontendKinds.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticFrontendKinds.td]: @MaskRay I removed the frontend warning, so that it follows how the driver handles unused flags. Now only `warning: argument unused during compilation` fires for `-x assembler`, and not for `-x assembler-with-cpp` to conform to most of other options.

File: clang/test/BoundsSafety/Frontend/ignored_for_ir.ll
Comment (rapidsna) [reply clang/test/BoundsSafety/Frontend/ignored_for_ir.ll]: @MaskRay fixed!

File: clang/test/Driver/bounds-safety-unused.c
Comment (rapidsna) [inline clang/test/Driver/bounds-safety-unused.c:9]: @nickdesaulniers The unused warning doesn't appear for `-x assembler-with-cpp` to follow the common behaviors. Instead, I added a note here that we need to add a targeted warning in future when assembler tries to use preprocessor directives to check if bounds safety ie enabled. WDYT?

File: clang/test/Driver/bounds-safety-unused.c
Comment (delcypher) [inline clang/test/Driver/bounds-safety-unused.c:2]: @rapidsna Maybe there should be a version of this part of the test that doesn't pass `-x <lang>` and instead passes to a `.s` file to make sure the right thing happens when in that case?

Comment (rapidsna) [reply clang/test/Driver/bounds-safety-unused.c:2]: I think `-x assembler` and `-x assembler-with-cpp` are more specific and should cover what I want. 

`.s` must be `-x assembler` (except for Darwin which does `-x assembler-with-cpp` for `.s`), but that specific behavior should be tested separately but not for this PR.

**Review by nickdesaulniers — APPROVED** [review]
If you make all of the newly added tests share a similar filename prefix, it will be easier to test via `llvm-lit -vv clang/test/Driver/bounds-safety*`.  `fbounds-safety.c` seems to break that convention.

File: clang/test/Driver/bounds-safety-unused.c
Comment (delcypher) [inline clang/test/Driver/bounds-safety-unused.c:10]: @rapidsna We should file an issue and refer to it in the code.

I've created #71206 for you and created a new `clang:bounds-safety` tag so we can track all the `-fbounds-safety` issues in open source.

Comment (rapidsna) [reply clang/test/Driver/bounds-safety-unused.c:10]: @delcypher Perfect, thanks!

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (MaskRay) [inline clang/lib/Frontend/CompilerInvocation.cpp:3621]: I think it's better to handle this language check in Driver, similar to `OPT_fminimize_whitespace` but for all `Inputs`.

 It's not necessary to mention 

> This is especially important because some build systems, like xcbuild and somewhat clumsy Makefiles, will pass
C_FLAGS to Clang while building assembly files.

since this is general and applies to other options. No need to emphasize it specifically for bounds checking.

**Review by MaskRay — CHANGES_REQUESTED** [review]

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (MaskRay) [inline clang/include/clang/Basic/DiagnosticGroups.td:1416]: This is `-Wignored-bounds-safety` is not used.

**Comment by MaskRay** [comment]
> [Driver][BoundsSafety] Add -fbounds-safety-experimental flag #70480

The patch implements `-fexperimental-bounds-safety`

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp:3621]: > I think it's better to handle this language check in Driver, similar to OPT_fminimize_whitespace but for all Inputs.

@MaskRay If we do it in Driver, wouldn't the option be silently ignored for unsupported languages when we invoke `clang -cc1` directly?

Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp:3621]: IMHO, that might be problematic because people would reasonably expect bounds-safety would work for C++/Obj-C/Obj-C++.

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (rapidsna) [reply clang/include/clang/Basic/DiagnosticGroups.td]: Removed it!

**Comment by rapidsna** [comment]
> > [Driver][BoundsSafety] Add -fbounds-safety-experimental flag #70480
> 
> The patch implements `-fexperimental-bounds-safety`

I just updated the title. I'll make sure I update the text when I squash the commits.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (delcypher) [reply clang/lib/Frontend/CompilerInvocation.cpp:3621]: There are various different options here.

1. The frontend silently accepts`-fexperimental-bounds-safety` with `-x c++` but doesn't actually enable `fexperimental-bounds-safety`.
2. The frontend accepts`-fexperimental-bounds-safety` with `-x c++` enables `-fexperimental-bounds-safety` and probably crashes because it's not supported.
3. The frontend rejects`-fexperimental-bounds-safety` with `-x c++` with an error.

I don't like (1.) because when writing frontend only tests it is inviting someone to accidentally write a test incorrectly because the frontend would never complain about the invalid configuration.

I don't like (2.) because crashing really isn't good when it could so easily be prevented.

I think we should do (3.) because it has none of the previously mentioned downsides.

Comment (delcypher) [reply clang/lib/Frontend/CompilerInvocation.cpp:3621]: To be clear. I think it's fine to also do the language check in the driver but I think the frontend should **also** do language check so that it is not possible to run it in an invalid configuration.

Comment (rapidsna) [reply clang/lib/Frontend/CompilerInvocation.cpp:3621]: > I think it's better to handle this language check in Driver, similar to `OPT_fminimize_whitespace` but for all `Inputs`.

@MaskRay Thanks for the review! I made it a driver error here: 
[f125532](https://github.com/llvm/llvm-project/pull/70480/commits/f125532235d5120027ec261204ae09d315d0fa14)

We also keep the -cc1 check in order to prevent it from running with an unsupported mode (we still need some checks in cc1 anyway in case we want to silently ignore the option, instead of triggering the broken mode).

**Comment by MaskRay** [comment]
It depends whether an option leads to a warning or is silently ignored for a language that is not implemented.
Many options are silently ignored. For example, -faligned-new and -fexceptions are silently ignored.

Sometimes, it is useful report a warning/error to prevent misuses.
Say, the bounds checking feature may be extended to C++ in the future and before that, we want to discourage C++ projects adopting the option.
Therefore, this patch proposes an error. I think this is fine.

However, implementing the checking in clang/lib/Driver is much more common.
clang/lib/Frontend/CompilerInvocation.cpp has some checking, but the majority is in clang/lib/Driver.

I am not convinced by your argument that you want to catch -cc1 misuses.
A user can use -Xclang xxx or -mllvm yyy to specify a lot of stuff and most are infeasible to verify in clang/lib/Frontend/CompilerInvocation.cpp.
If a user is motivated to bypass the compiler' error checking, there is not much the compiler can do.

**Comment by rapidsna** [comment]
> However, implementing the checking in clang/lib/Driver is much more common. clang/lib/Frontend/CompilerInvocation.cpp has some checking, but the majority is in clang/lib/Driver.

@MaskRay To be clear, the check is now in Driver as you suggested. It's just that the frontend also has some extra checks too. So, you want me to remove the extra checks in the frontend?

**Comment by rapidsna** [comment]
> Sometimes, it is actually useful to expose some experimental features that we don't feel comfortable surfacing as driver options as cc1 options. The cc1 options give the users to experiment with the feature. It's the users's responsibility to adapt when the feature changes.

Ok. I'll fix this.

**Comment by MaskRay** [comment]
> @MaskRay To be clear, the check is now in Driver as you suggested. It's just that the frontend also has some extra checks too. So, you want me to remove the extra checks in the frontend?

Yes, otherwise it's duplicated check.

In addition, I think our convention is to add the option when the feature is ready, rather than add a no-op option, than build functional patches on top of it.

**Review by MaskRay — CHANGES_REQUESTED** [review]
> -fbounds-safety-experimental is an experimental flag for -fbounds-safety,

-fexperimental-bounds-safety

**Comment by rapidsna** [comment]
> > -fbounds-safety-experimental is an experimental flag for -fbounds-safety,
> 
> -fexperimental-bounds-safety

Changed the description in the PR! I'll adjust the commit message too when I squash all the changes once I get your approval.

**Comment by rapidsna** [comment]
> > @MaskRay To be clear, the check is now in Driver as you suggested. It's just that the frontend also has some extra checks too. So, you want me to remove the extra checks in the frontend?
> 
> Yes, otherwise it's duplicated check.

I just removed the check in the frontend!

**Comment by rapidsna** [comment]
> In addition, I think our convention is to add the option when the feature is ready, rather than add a no-op option, than build functional patches on top of it.

@MaskRay  This feature will involve a lot of incremental patches. And we will still need an experimental flag to test the incremental functionalities that are added. I can make it a CC1 only flag for now. Would it work for you better?

**Comment by MaskRay** [comment]
> > In addition, I think our convention is to add the option when the feature is ready, rather than add a no-op option, than build functional patches on top of it.
> 
> @MaskRay This feature will involve a lot of incremental patches. And we will still need an experimental flag to test the incremental functionalities that we will be adding. I can make it a CC1 only flag for now. Would it work for you better?

Yes. This sound good to me.

You will need to remove the lib/Driver change and the language compatibility check, then.

I think the patch should still be held off until the first patch that does some basic functionality is ready.

**Comment by bwendling** [comment]
@rapidsna, what's the status of this PR? If it's bit-rotted, I can try to take over.

**Comment by bwendling** [comment]
Also, I think there was a PR that added delayed parsing of attributes? Is that correct, or a false memory? :-)

**Comment by rapidsna** [comment]
@bwendling We've already landed the flag in https://github.com/llvm/llvm-project/pull/92623. I will close this one.

**Comment by rapidsna** [comment]
> Also, I think there was a PR that added delayed parsing of attributes? Is that correct, or a false memory? :-)

@bwendling That too has already been merged in https://github.com/llvm/llvm-project/pull/88596

**Pull request closed without merging**
